app/src/main/java/org/mozilla/focus/home/HomeFragment.java

```diff
@@ -201,29 +201,67 @@ protected void onPostExecute(String line) {
     }
 
     private void setUpBanner(Context context) {
+        // 1. Read from cache first
+        String[] fromCache = readFromCache(context);
+        if (fromCache.length != 0) {
+            setUpBannerFromConfig(fromCache);
+        }
+        // 2. Load item for next initialization
         String manifest = AppConfigWrapper.getBannerRootConfig(context);
         if (TextUtils.isEmpty(manifest)) {
-            return;
-        }
-        // Not using a local variable to prevent reference to be cleared before returning.
-        onRootConfigLoadedListener = configArray -> {
-            try {
-                if (configArray == null) {
+            deleteCache(context);
+            banner.setAdapter(null);
+            showBanner(false);
+        } else {
+            // Not using a local variable to prevent reference to be cleared before returning.
+            onRootConfigLoadedListener = configArray -> {
+                if (Arrays.equals(fromCache, configArray)) {
                     return;
                 }
-                BannerAdapter bannerAdapter = new BannerAdapter(configArray, arg -> FragmentListener.notifyParent(this, FragmentListener.TYPE.OPEN_URL_IN_NEW_TAB, arg));
-                banner.setAdapter(bannerAdapter);
-                banner.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.HORIZONTAL, false));
-                SnapHelper snapHelper = new PagerSnapHelper();
-                snapHelper.attachToRecyclerView(banner);
-                showBanner(true);
-            } catch (JSONException e) {
-                Logger.throwOrWarn(TAG, "Invalid Config: " + e.getMessage());
-            } finally {
+                writeToCache(context, configArray);
```

Comment:
It seems we are doing File IO on the main thread. Maybe we can file another PR to move this into LiveData pattern?

Reply:
I think this is very lightweight I/O so should be okay. But I'll file a bug.